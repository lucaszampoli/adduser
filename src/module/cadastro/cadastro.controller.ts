import { Controller, Post, Get, Res, HttpStatus, Body, Param, NotFoundException, HttpCode, Req } from '@nestjs/common';
import { Response, Request } from 'express';
import { HttpResponse } from '../../utils/http.response';
import { Status } from '../../utils/status.entity';
import { CadastroService } from './cadastro.service';
import { UserArgs, UserLogin } from './repositories/cadastro.entity';

@Controller('cadastro')
export class CadastroController {

    constructor(private readonly service: CadastroService) { }

    @Post('signup')
    public async registrar(@Res() res: Response, @Body() filter: UserArgs): Promise<Response> {
        let data;
        try {
            data = await this.service.registrar(filter);
            return res.status(HttpStatus.OK).json(
                new HttpResponse(new Status(HttpStatus.OK), data));
        } catch (err) {
            switch(err.code) {
                case 11000:
                    return res.status(HttpStatus.INTERNAL_SERVER_ERROR).json(
                        new HttpResponse(new Status(HttpStatus.INTERNAL_SERVER_ERROR, 'E-mail já existente.')));
                default:
                    return res.status(HttpStatus.INTERNAL_SERVER_ERROR).json(
                        new HttpResponse(new Status(HttpStatus.INTERNAL_SERVER_ERROR, err.errmsg)));
            }
        }
    }

    @Post('signin')
    public async logar(@Res() res: Response, @Body() filter: UserLogin): Promise<Response> {
        let data;
        try {
            data = await this.service.logar(filter);
            return res.status(HttpStatus.OK).json(
                new HttpResponse(new Status(HttpStatus.OK), data));
        } catch (err) {
            switch(err.code) {
                case 401:
                    return res.status(HttpStatus.UNAUTHORIZED).json(
                        new HttpResponse(new Status(HttpStatus.UNAUTHORIZED, err.errmsg)));
                default:
                    return res.status(HttpStatus.INTERNAL_SERVER_ERROR).json(
                        new HttpResponse(new Status(HttpStatus.INTERNAL_SERVER_ERROR, err.errmsg)));
            }
        }
    }

    @Get('search/:userid')
    public async buscar(@Req() req: Request, @Res() res: Response, @Param() params): Promise<Response> {
        let data;
        try {
            const authorization = req.get('Authorization');
            if(!authorization) {
                return res.status(HttpStatus.UNAUTHORIZED).json(
                    new HttpResponse(new Status(HttpStatus.UNAUTHORIZED, 'Não autorizado')));
            }
            data = await this.service.buscar({...params, token: authorization});
            return res.status(HttpStatus.OK).json(
                new HttpResponse(new Status(0), data));
        } catch (err) {
            switch(err.code) {
                case 401:
                    return res.status(HttpStatus.UNAUTHORIZED).json(
                        new HttpResponse(new Status(HttpStatus.UNAUTHORIZED, err.errmsg)));
                default:
                    return res.status(HttpStatus.INTERNAL_SERVER_ERROR).json(
                        new HttpResponse(new Status(HttpStatus.INTERNAL_SERVER_ERROR, err.errmsg)));
            }
        }
    }

}