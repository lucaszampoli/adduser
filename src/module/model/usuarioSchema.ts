import { model, Schema } from "mongoose";

const UsuarioSchema = new Schema(
  {
    userid: { type: String, required: true, index: true, unique: true },
    nome: { type: String, required: true },
    email: { type: String, required: true, index: true, unique: true },
    senha: { type: String, required: true },
    telefones: { type: Array },
    data_criacao: { type: Date, default: Date.now },
    data_atualizacao: { type: Date, default: Date.now },
    ultimo_login: { type: Date, default: Date.now },
    token: { type: String },
  },
  {
    strict: true,
    collection: "usuarios"
  }
);

UsuarioSchema.pre("save", next => {
  const now = new Date();
  if (!this.data_criacao) {
    this.data_criacao = now;
  }
  this.data_atualizacao = now;
  next();
});

export const usuarioModel = model("usuarios", UsuarioSchema, "usuarios");
