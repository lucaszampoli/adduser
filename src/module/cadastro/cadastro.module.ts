import { Module, MiddlewareConsumer } from '@nestjs/common';
import { CadastroController } from './cadastro.controller';
import { CadastroService } from './cadastro.service';
import { JwtModule } from '@nestjs/jwt';

@Module({
    imports: [JwtModule.register({
        secret: 'YWNob3UgcXVlIGV1IGVzdGF2YSBicmluY2FuZG8gPwo='
    })],
    controllers: [CadastroController],
    providers: [CadastroService],
    exports: [CadastroService],
})
export class CadastroModule { }