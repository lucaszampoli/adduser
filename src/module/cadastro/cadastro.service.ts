import { NotFoundException, HttpStatus, Injectable } from '@nestjs/common';
import { IsString } from 'class-validator';
import { UserArgs, UserLogin, UserSearch } from './repositories/cadastro.entity';
import { usuarioModel } from '../model/usuarioSchema';
import { JwtService } from  '@nestjs/jwt';
import * as crypto from "crypto";
import * as moment from "moment";

@Injectable()
export class CadastroService {

    private jwtService: JwtService;

    constructor(jwtService: JwtService) {
        this.jwtService = jwtService;
    }

    private async getNewId() {
        try {
            const usuarios: any = await usuarioModel.find({}).sort({ _id: -1 }).limit(1);
            if (usuarios.length > 0) {
                const seq = usuarios? parseFloat(usuarios[0].userid) + 1 : 1;
                return `0000${seq}`.slice(-5);
            } else {
                return `00001`;
            }
        } catch (err) {
            console.log(err);
            return '';
        }
    }

    public async registrar(filter: UserArgs): Promise<any> | never {
        try {
            const now = new Date();
            const payload = `${filter.nome}${filter.email}${now}`
            const user: any = await new usuarioModel({
                userid: await this.getNewId(),
                nome: filter.nome,
                email: filter.email,
                senha: crypto.createHmac('sha512', filter.senha).digest('hex'),
                telefones: filter.telefones,
                token: this.jwtService.sign(payload),
            }).save();
            return Promise.resolve({
                'userId': user.userid,
                'data_criacao': user.data_criacao,
                'data_atualizacao': user.data_atualizacao,
                'ultimo_login': user.ultimo_login,
                'token': user.token,
            });
        } catch (err) {
            throw err
        }
    }

    public async logar(filter: UserLogin): Promise<any> | never {
        try {
            const user: any = await usuarioModel.find({'email': { $regex: filter.email }} ).sort({ _id: -1 }).limit(1);
            if (user[0].senha === crypto.createHmac('sha512', filter.senha).digest('hex')) {
                const now = new Date();
                const payload = `${user.nome}${user.email}${now}`
                const token = this.jwtService.sign(payload);
                const updateUser = await usuarioModel.findOneAndUpdate({ userId: user[0].userId }, { ultimo_login: now, token: token }, { new: true })
                return Promise.resolve({
                    'userId': updateUser.userid,
                    'data_criacao': updateUser.data_criacao,
                    'data_atualizacao': updateUser.data_atualizacao,
                    'ultimo_login': updateUser.ultimo_login,
                    'token': updateUser.token,
                });
            } else {
                const errObj = {
                    code: HttpStatus.UNAUTHORIZED,
                    errmsg: 'Usuário e/ou senha inválidos',
                }
                return Promise.reject(errObj);
            } 
        } catch (err) {
            throw err;
        }
    }

    public async buscar(filter: UserSearch): Promise<any> | never {
        try {
            const user: any = await usuarioModel.find({'userid': { $regex: filter.userid }} ).sort({ _id: -1 }).limit(1);
            if(user.length > 0) {
                if (`Bearer {${user[0].token}}` === filter.token) {
                    const ultimo_login = moment(user[0].ultimo_login).format('YYYYMMDDhhmmss');
                    const hora_atual = moment().subtract(30, 'minutes').format('YYYYMMDDhhmmss');
                    if (parseInt(hora_atual) > parseInt(ultimo_login)) {
                        const errObj = {
                            code: HttpStatus.UNAUTHORIZED,
                            errmsg: 'Sessão inválida',
                        }
                        return Promise.reject(errObj);
                    } else {
                        return Promise.resolve({
                            'userId': user[0].userid,
                            'data_criacao': user[0].data_criacao,
                            'data_atualizacao': user[0].data_atualizacao,
                            'ultimo_login': user[0].ultimo_login,
                            'token': user[0].token,
                        });
                    }
                } else {
                    const errObj = {
                        code: HttpStatus.UNAUTHORIZED,
                        errmsg: 'Não autorizado',
                    }
                    return Promise.reject(errObj);
                }
            } else {
                const errObj = {
                    code: HttpStatus.INTERNAL_SERVER_ERROR,
                    errmsg: 'Usuário inexistente',
                }
                return Promise.reject(errObj);
            }
        } catch (err) {
            throw err;
        }
    }

}