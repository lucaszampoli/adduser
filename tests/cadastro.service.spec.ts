import { CadastroService } from '../src/module/cadastro/cadastro.service';
import { JwtService } from '@nestjs/jwt';
import mockingoose from 'mockingoose';
import { usuarioModel } from '../src/module/model/usuarioSchema';
import { UserArgs, UserLogin, UserSearch } from '../src/module/cadastro/repositories/cadastro.entity';
import * as moment from 'moment';

const mockJwt = {
    sign: () => {}
};
const mockUserArgs = {
    "nome": "Epaminondas",
    "email": "epaminondas@gmail.com",
    "senha": "password",
    "telefones": [
        {
            "numero": "123456789",
            "ddd": "11"
        },
        {
            "numero": "987654321",
            "ddd": "12"
        }
    ]
} as UserArgs;
const mockUserLogin = {
	"email": "epaminondas@gmail.com",
	"senha": "password"
} as UserLogin;
const mockParams = {
    userid: '123',
    token: 'Bearer {test}'
} as UserSearch;
const mockUser = {
    "userid": "123",
    "nome": "Epaminondas",
    "email": "epaminondas@gmail.com",
    "senha": "ded1bda29058c02f011c01087e12543684821c54d42ac2a9924a1a0ee10830b5a2028a90febc0cda5188517bc79aecdbd720b00895181571496e8688c57ddfa0",
    "data_criacao": new Date(),
    "data_atualizacao": new Date(),
    "ultimo_login": new Date(),
    "token": "test"
}

const finderMock = query => {
    return [mockUser];
};
const saveMock = query => {
    return [mockUser];
};

describe('Cadastro Service', () => {
    let service: CadastroService;
    let jwtService: JwtService;

    beforeEach(() => {
        jwtService = mockJwt as any;
        service = new CadastroService(jwtService);
        mockingoose(usuarioModel).toReturn(saveMock, 'save');
        mockingoose(usuarioModel).toReturn(finderMock, 'find');
        mockingoose(usuarioModel).toReturn(finderMock, 'findOneAndUpdate');
    });

    describe('List operation cases', () => {
        it('should have instance service', () => {
            expect(service).toBeTruthy();
        });

        it('should test registrar', async () => {
            expect(await service.registrar(mockUserArgs)).toBeTruthy();
        });

        it('should test logar', async () => {
            expect(await service.logar(mockUserLogin)).toBeTruthy();
        });

        it('should test buscar', async () => {
            expect(await service.buscar(mockParams)).toBeTruthy();
        });

    })

})