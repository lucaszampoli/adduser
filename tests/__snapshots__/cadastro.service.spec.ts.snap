// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cadastro Service List operation cases should test logar: find query 1`] = `
Object {
  "email": Object {
    "$regex": "epaminondas@gmail.com",
  },
}
`;

exports[`Cadastro Service List operation cases should test registrar: find query 1`] = `Object {}`;
